packages/data/src/components/use-dispatch/use-dispatch-with-map.js

```diff
@@ -0,0 +1,37 @@
+/**
+ * External dependencies
+ */
+import { mapValues } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { useMemo } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import useRegistry from '../registry-provider/use-registry';
+
+function proxyDispatch( propName, registry, ...args ) {
+	this( registry.dispatch, registry )[ propName ]( ...args );
+}
+
+const useDispatchWithMap = ( dispatchMap ) => {
+	const registry = useRegistry();
+	const currentDispatchProps = dispatchMap( registry.dispatch, registry );
+	return useMemo( () => mapValues(
+		currentDispatchProps,
+		( dispatcher, propName ) => {
+			if ( typeof dispatcher !== 'function' ) {
+				// eslint-disable-next-line no-console
+				console.warn(
+					`Property ${ propName } returned from dispatchMap in useDispatchWithMap must be a function.`
+				);
+			}
+			return proxyDispatch.bind( dispatchMap, propName, registry );
```

Comment:
You could also just declare an arrow function here to avoid the `this` complexity.

Reply:
I just ended up returning the `dispatcher` variable instead (see 7d88dd1)

Reply:
You still need to call `mapDispatch` again when a `dispatcher` is called, so the selected state is fresh.

Reply:
so something like this instead?

```js
return ( ...args ) => dispatcher( registry.dispatch, registry )[ propName ]( ...args )
```

Reply:
Exactly 👍

Reply:
actually I had to implement `useRef` to grab the latest dispatchMap (see 5856f20)  Some failing e2e tests revealed that the above example I tried was not sufficient.

Reply:
See https://github.com/WordPress/gutenberg/pull/15896#discussion_r288836476